source/core/scheduler.cpp

```diff
@@ -42,15 +50,19 @@ void Scheduler::_run() {
   _interrupt = false;
 
   while (!_interrupt) {
-    Task task;
-    if (_taskQueue.pop(task)) {
-      // there was something in the queue, ececute the task
-      task();
-
-    } else {
-      // the queue is empty, sleep for a bit
-      std::this_thread::sleep_for(_sleepMilliseconds);
+    std::unique_lock<std::mutex> lock(_mutex);
+    while (_taskQueue.empty()) {
+      // wait while there is nothing in the queue, guards for spurious wakeups.
+      _conditionVariable.wait(lock);
     }
+    // make sure the mutex is locked and pop an element from the queue (which is
+    // guaranteed to not be empty)
+    if (!lock) lock.lock();
```

Comment:
You make these checks here (and in other places), but one of the properties of the condition variable is that the mutex is reacquired upon wakeup. As such I think this check is unnecessary?